src/NuGet.Core/NuGet.Commands/RestoreCommand/DependencyGraphResolver.cs

```diff
@@ -1389,307 +859,562 @@ private static bool VersionRangePreciseEquals(VersionRange a, VersionRange b)
+                LibraryRangeIndex[] currentDependencyGraphItemPath = currentDependencyGraphItem.Path;
+                HashSet<LibraryDependencyIndex>? currentDependencyGraphItemSuppressions = currentDependencyGraphItem.Suppressions;
+                bool currentDependencyGraphItemIsDirectPackageReferenceFromRootProject = currentDependencyGraphItem.IsDirectPackageReferenceFromRootProject;
 
-            public LibraryDependencyIndex Intern(LibraryDependency libraryDependency)
-            {
-                lock (_lockObject)
+                // Determine if what is being processed is the root project itself which has different rules vs a transitive dependency
+                bool isRootProject = currentDependencyGraphItem.LibraryDependencyIndex == LibraryDependencyIndex.Project;
+
+                GraphItem<RemoteResolveResult> currentGraphItem = await currentDependencyGraphItem.GetGraphItemAsync(_request, projectTargetFramework?.PackagesToPrune, isRootProject, context, _logger);
+
+                LibraryDependencyTarget typeConstraint = currentLibraryDependency.LibraryRange.TypeConstraint;
+                if (evictions.TryGetValue(currentLibraryRangeIndex, out (LibraryRangeIndex[], LibraryDependencyIndex, LibraryDependencyTarget) eviction))
                 {
-                    string key = libraryDependency.Name;
-                    if (!_table.TryGetValue(key, out LibraryDependencyIndex index))
+                    (LibraryRangeIndex[] evictedPath, LibraryDependencyIndex evictedDepIndex, LibraryDependencyTarget evictedTypeConstraint) = eviction;
+
+                    // If we evicted this same version previously, but the type constraint of currentRef is more stringent (package), then do not skip the current item - this is the one we want.
+                    // This is tricky. I don't really know what this means. Normally we'd key off of versions instead.
+                    if (!((evictedTypeConstraint == LibraryDependencyTarget.PackageProjectExternal || evictedTypeConstraint == LibraryDependencyTarget.ExternalProject) &&
+                        currentLibraryDependency.LibraryRange.TypeConstraint == LibraryDependencyTarget.Package))
                     {
-                        index = (LibraryDependencyIndex)_nextIndex++;
-                        _table.TryAdd(key, index);
+                        continue;
                     }
-
-                    return index;
                 }
-            }
 
-            public LibraryDependencyIndex Intern(CentralPackageVersion centralPackageVersion)
-            {
-                string key = centralPackageVersion.Name;
-                if (!_table.TryGetValue(key, out LibraryDependencyIndex index))
+                // Determine if a dependency with the same name has not already been chosen
+                if (!resolvedDependencyGraphItems.TryGetValue(currentLibraryDependencyIndex, out ResolvedDependencyGraphItem? chosenResolvedItem))
```

Comment:
I feel like current and chosen aren't the best names for these types. 
Current is the one that's being processed and chosen is the existing selection. 

Sometimes I find myself thinking, "current" is the one "currently chosen". 

Maybe it's just me, but if you have similar confusions, maybe wqe can try to figure out an alternative.

Reply:
I like selected and current. 

We can always keep the current naming in this PR and thne just do a quick follow-up after we've talked.

I'll resolve this for now since I don't want to create more churn for this PR.